File: apps/desktop/services.mjs
import { randomUUID } from 'node:crypto';
import { join } from 'node:path';
import {
  openWorkspace, createSession, getSession, listSessions, appendEvents,
  listEvents, replaceFindings, listFindings
} from '../../src/core/store.mjs';
import { inferFindings } from '../../src/core/infer.mjs';
import { compareSessions } from '../../src/core/compare.mjs';
import { exportSession } from '../../src/core/exporter.mjs';
import { importHar as readHar } from '../../src/importers/har.mjs';
import { importProcmon as readProcmon } from '../../src/importers/procmon.mjs';
import { importJsonl as readJsonl } from '../../src/importers/jsonl.mjs';
import { launchTarget } from '../../src/observe/launch.mjs';
import { listCollectorManifests, describeCollector } from '../../src/collectors/registry.mjs';
import { resolveNativeCollectorBinary } from '../../src/collectors/native-sidecar.mjs';
import { runExternalCollector } from '../../src/collectors/host.mjs';
import { DesktopError } from './errors.mjs';
import {
  assertCollectorInstanceId, assertSessionId, validateCollectorStartRequest,
  validateDialogFilters, validateLaunchRequest, validateSessionInput
} from './validators.mjs';

function withDb(dbPath, fn) {
  const db = openWorkspace(dbPath);
  try { return fn(db); } finally { db.close(); }
}

async function withDbAsync(dbPath, fn) {
  const db = openWorkspace(dbPath);
  try { return await fn(db); } finally { db.close(); }
}

function requireSession(db, sessionId) {
  assertSessionId(sessionId);
  const session = getSession(db, sessionId);
  if (!session) throw new DesktopError('SESSION_NOT_FOUND', 'The selected session no longer exists.');
  return session;
}

function importResult(stored) {
  return { cancelled: false, imported: stored.length, kinds: [...new Set(stored.map((event) => event.kind))].sort() };
}

function wrapImport(error) {
  if (error instanceof DesktopError) throw error;
  throw new DesktopError('IMPORT_INVALID_FILE', 'The selected evidence file could not be imported.');
}

function safeName(value) {
  return String(value || 'session').replace(/[^A-Za-z0-9._-]+/g, '-').replace(/^-+|-+$/g, '').slice(0, 80) || 'session';
}

function collectorSnapshot(record) {
  return {
    instanceId: record.instanceId,
    collectorId: record.collectorId,
    sessionId: record.sessionId,
    status: record.status,
    diagnostics: [...record.diagnostics],
    code: record.code ?? null
  };
}

function collectorError(error, fallback = 'COLLECTOR_RUNTIME_FAILED') {
  const code = typeof error?.code === 'string' ? error.code : fallback;
  return new DesktopError(code, code === 'COLLECTOR_NOT_FOUND'
    ? 'The requested collector executable is not installed.'
    : 'The collector operation could not be completed.');
}

export function createDesktopServices({ dbPath, dialogs, appInfo, platform = process.platform }) {
  if (typeof dbPath !== 'string' || !dbPath) throw new TypeError('dbPath is required');
  const nativeDialogs = dialogs ?? {};
  const metadata = { name: appInfo?.name ?? 'HarnessScope', version: appInfo?.version ?? '0.0.0' };
  const collectorInstances = new Map();

  const services = {
    async appInfo() {
      return { ...metadata, platform };
    },

    async workspaceInfo() {
      return { dbPath };
    },

    async sessionList() {
      return withDb(dbPath, (db) => listSessions(db));
    },

    async sessionCreate(input) {
      const value = validateSessionInput(input);
      return withDb(dbPath, (db) => createSession(db, value.name, value.mode));
    },

    async timelineGet(sessionId) {
      return withDb(dbPath, (db) => {
        const session = requireSession(db, sessionId);
        return { session, events: listEvents(db, sessionId), findings: listFindings(db, sessionId) };
      });
    },

    async inferenceRun(sessionId) {
      return withDb(dbPath, (db) => {
        const session = requireSession(db, sessionId);
        const findings = inferFindings(listEvents(db, sessionId));
        replaceFindings(db, sessionId, findings);
        return { session, findings: listFindings(db, sessionId) };
      });
    },

    async compareRun(sessionA, sessionB) {
      return withDb(dbPath, (db) => {
        const load = (id) => {
          const session = requireSession(db, id);
          return { session, events: listEvents(db, id), findings: listFindings(db, id) };
        };
        return compareSessions(load(sessionA), load(sessionB));
      });
    },

    async importHar(sessionId, filePath = null) {
      assertSessionId(sessionId);
      const selected = filePath ?? await nativeDialogs.pickFile?.({ filters: [{ name: 'HTTP Archive', extensions: ['har'] }] });
      if (!selected) return { cancelled: true };
      try {
        return withDb(dbPath, (db) => {
          requireSession(db, sessionId);
          return importResult(appendEvents(db, sessionId, readHar(selected)));
        });
      } catch (error) { return wrapImport(error); }
    },

    async importProcmon(sessionId, filePath = null) {
      assertSessionId(sessionId);
      const selected = filePath ?? await nativeDialogs.pickFile?.({ filters: [{ name: 'Procmon CSV', extensions: ['csv'] }] });
      if (!selected) return { cancelled: true };
      try {
        return withDb(dbPath, (db) => {
          requireSession(db, sessionId);
          return importResult(appendEvents(db, sessionId, readProcmon(selected)));
        });
      } catch (error) { return wrapImport(error); }
    },

    async importJsonl(sessionId, filePath = null, mapPath = null) {
      assertSessionId(sessionId);
      const selected = filePath ?? await nativeDialogs.pickFile?.({ filters: [{ name: 'JSON Lines', extensions: ['jsonl', 'ndjson'] }] });
      if (!selected) return { cancelled: true };
      const selectedMap = mapPath ?? await nativeDialogs.pickFile?.({ filters: [{ name: 'YAML mapping', extensions: ['yaml', 'yml'] }] });
      if (!selectedMap) return { cancelled: true };
      try {
        return withDb(dbPath, (db) => {
          requireSession(db, sessionId);
          return importResult(appendEvents(db, sessionId, readJsonl(selected, selectedMap)));
        });
      } catch (error) { return wrapImport(error); }
    },

    async launchRun(sessionId, request) {
      assertSessionId(sessionId);
      const launch = validateLaunchRequest(request);
      return withDbAsync(dbPath, async (db) => {
        requireSession(db, sessionId);
        try {
          return await launchTarget({ db, sessionId, target: launch.target, args: launch.args, cwd: launch.cwd ?? process.cwd() });
        } catch {
          throw new DesktopError('LAUNCH_FAILED', 'The selected command could not be launched.');
        }
      });
    },

    async exportRun(sessionId, outDir = null) {
      assertSessionId(sessionId);
      const selected = outDir ?? await nativeDialogs.pickDirectory?.();
      if (!selected) return { cancelled: true };
      return withDb(dbPath, (db) => {
        const session = requireSession(db, sessionId);
        try {
          const target = outDir ? selected : join(selected, `HarnessScope-${safeName(session.name)}`);
          return { cancelled: false, ...exportSession({ db, sessionId, outDir: target }) };
        } catch {
          throw new DesktopError('EXPORT_FAILED', 'The behavioral spec could not be exported.');
        }
      });
    },

    async collectorList() {
      return listCollectorManifests({ platform }).map((manifest) => {
        let available = true;
        try { resolveNativeCollectorBinary({ platform }); }
        catch { available = false; }
        return { ...manifest, available };
      });
    },

    async collectorDescribe(collectorId) {
      if (typeof collectorId !== 'string' || !collectorId.trim()) {
        throw new DesktopError('INVALID_COLLECTOR_REQUEST', 'Collector id is invalid.');
      }
      const manifest = describeCollector(collectorId.trim(), { platform });
      if (!manifest) throw new DesktopError('COLLECTOR_NOT_FOUND', 'The requested collector is unavailable on this platform.');
      let available = true;
      try { resolveNativeCollectorBinary({ platform }); }
      catch { available = false; }
      return { ...manifest, available };
    },

    async collectorStart(sessionId, input) {
      assertSessionId(sessionId);
      const requestValue = validateCollectorStartRequest(input);
      withDb(dbPath, (db) => requireSession(db, sessionId));

      const manifest = requestValue.collectorId
        ? describeCollector(requestValue.collectorId, { platform })
        : null;
      if (requestValue.collectorId && !manifest) {
        throw new DesktopError('COLLECTOR_NOT_FOUND', 'The requested collector is unavailable on this platform.');
      }

      const instanceId = randomUUID();
      let command;
      try {
        command = requestValue.command ?? resolveNativeCollectorBinary({ platform });
      } catch (error) {
        throw collectorError(error, 'COLLECTOR_NOT_FOUND');
      }

      const protocolRequest = {
        sdkVersion: '1',
        instanceId,
        requestedCapabilities: manifest?.capabilities ? [...manifest.capabilities] : [],
        paths: [...requestValue.paths],
        hashFiles: requestValue.hashFiles,
        target: { ...requestValue.target, args: [...requestValue.target.args] }
      };
      if (manifest) protocolRequest.collectorId = manifest.id;

      const controller = new AbortController();
      const record = {
        instanceId,
        collectorId: manifest?.id ?? requestValue.collectorId ?? 'external.collector',
        sessionId,
        status: 'starting',
        diagnostics: [],
        code: null,
        stopRequested: false,
        controller,
        promise: null
      };
      collectorInstances.set(instanceId, record);

      const db = openWorkspace(dbPath);
      try {
        record.promise = runExternalCollector({
          db,
          sessionId,
          command,
          args: [],
          request: protocolRequest,
          expectedCollectorId: manifest?.id,
          signal: controller.signal,
          onStatus(status) { record.status = status; }
        }).then((result) => {
          record.status = result.status;
          record.collectorId = result.collectorId;
          record.diagnostics = [...result.diagnostics];
          return result;
        }).catch((error) => {
          record.code = typeof error?.code === 'string' ? error.code : 'COLLECTOR_RUNTIME_FAILED';
          record.status = record.stopRequested ? 'stopped' : 'failed';
          return null;
        }).finally(() => {
          db.close();
        });
      } catch (error) {
        db.close();
        collectorInstances.delete(instanceId);
        throw collectorError(error);
      }

      return collectorSnapshot(record);
    },

    async collectorStop(instanceId) {
      assertCollectorInstanceId(instanceId);
      const record = collectorInstances.get(instanceId);
      if (!record) throw new DesktopError('COLLECTOR_NOT_FOUND', 'The collector instance does not exist.');
      if (record.status === 'stopped' || record.status === 'failed') return collectorSnapshot(record);
      record.stopRequested = true;
      record.status = 'stopping';
      record.controller.abort();
      return collectorSnapshot(record);
    },

    async collectorStatus(instanceId) {
      assertCollectorInstanceId(instanceId);
      const record = collectorInstances.get(instanceId);
      if (!record) throw new DesktopError('COLLECTOR_NOT_FOUND', 'The collector instance does not exist.');
      return collectorSnapshot(record);
    },

    async collectorShutdown() {
      const waits = [];
      for (const record of collectorInstances.values()) {
        if (record.status !== 'stopped' && record.status !== 'failed') {
          record.stopRequested = true;
          record.status = 'stopping';
          record.controller.abort();
        }
        if (record.promise) waits.push(record.promise);
      }
      await Promise.allSettled(waits);
    },

    async pickDirectory() {
      return await nativeDialogs.pickDirectory?.() ?? null;
    },

    async pickFile(filters) {
      return await nativeDialogs.pickFile?.({ filters: validateDialogFilters(filters) }) ?? null;
    }
  };

  return services;
}

File: crates/harnesscope-core/src/services.rs
use crate::{
    CompareResult, CoreError, ExportResult, LaunchRequest, LaunchResult, Session, SessionSnapshot,
    TraceEvent, TraceEventInput, Workspace, WorkspaceLock, compare_sessions, export_session,
    import_har, import_jsonl, import_procmon, infer_findings, launch_target,
};
use serde::{Deserialize, Serialize};
use std::{collections::BTreeSet, fmt, path::Path, path::PathBuf};
use time::{OffsetDateTime, macros::format_description};

#[derive(Clone, Debug, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct AppInfo {
    pub name: String,
    pub version: String,
    pub platform: String,
}

#[derive(Clone, Debug, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct WorkspaceInfo {
    pub db_path: PathBuf,
}

#[derive(Clone, Debug, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct ImportResult {
    pub cancelled: bool,
    pub imported: usize,
    pub kinds: Vec<String>,
}

#[derive(Clone, Debug, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct InferenceResult {
    pub session: Session,
    pub findings: Vec<crate::Finding>,
}

#[derive(Clone, Debug, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct ServiceExportResult {
    pub cancelled: bool,
    pub out_dir: PathBuf,
    pub files: Vec<String>,
}

pub struct AppServices {
    workspace: Workspace,
    _lease: WorkspaceLock,
    name: String,
    version: String,
    platform: String,
}

impl fmt::Debug for AppServices {
    fn fmt(&self, formatter: &mut fmt::Formatter<'_>) -> fmt::Result {
        formatter
            .debug_struct("AppServices")
            .field("db_path", &self.workspace.path())
            .field("name", &self.name)
            .field("version", &self.version)
            .field("platform", &self.platform)
            .finish_non_exhaustive()
    }
}

fn today_utc() -> Result<String, CoreError> {
    OffsetDateTime::now_utc()
        .format(&format_description!("[year]-[month]-[day]"))
        .map_err(|error| CoreError::Time(error.to_string()))
}

fn import_result(events: Vec<crate::TraceEvent>) -> ImportResult {
    let kinds = events
        .iter()
        .map(|event| event.kind.clone())
        .collect::<BTreeSet<_>>()
        .into_iter()
        .collect();
    ImportResult {
        cancelled: false,
        imported: events.len(),
        kinds,
    }
}

impl AppServices {
    pub fn open(
        db_path: impl AsRef<Path>,
        name: &str,
        version: &str,
        runtime: &str,
    ) -> Result<Self, CoreError> {
        let db_path = db_path.as_ref();
        if name.trim().is_empty() || version.trim().is_empty() || runtime.trim().is_empty() {
            return Err(CoreError::InvalidArgument);
        }
        let lease = WorkspaceLock::acquire(db_path, runtime)?;
        let workspace = Workspace::open(db_path)?;
        Ok(Self {
            workspace,
            _lease: lease,
            name: name.to_owned(),
            version: version.to_owned(),
            platform: std::env::consts::OS.to_owned(),
        })
    }

    pub fn app_info(&self) -> AppInfo {
        AppInfo {
            name: self.name.clone(),
            version: self.version.clone(),
            platform: self.platform.clone(),
        }
    }

    pub fn workspace_info(&self) -> WorkspaceInfo {
        WorkspaceInfo {
            db_path: self.workspace.path().to_path_buf(),
        }
    }

    fn require_session(&self, session_id: &str) -> Result<Session, CoreError> {
        if session_id.trim().is_empty() {
            return Err(CoreError::InvalidArgument);
        }
        self.workspace
            .get_session(session_id)?
            .ok_or(CoreError::SessionNotFound)
    }

    pub fn session_list(&self) -> Result<Vec<Session>, CoreError> {
        self.workspace.list_sessions()
    }

    pub fn session_create(&self, name: &str, mode: &str) -> Result<Session, CoreError> {
        if name.trim().is_empty() || mode.trim().is_empty() {
            return Err(CoreError::InvalidArgument);
        }
        self.workspace.create_session(name, mode)
    }

    pub fn timeline_get(&self, session_id: &str) -> Result<SessionSnapshot, CoreError> {
        let session = self.require_session(session_id)?;
        Ok(SessionSnapshot {
            session,
            events: self.workspace.list_events(session_id)?,
            findings: self.workspace.list_findings(session_id)?,
        })
    }

    pub fn inference_run(&self, session_id: &str) -> Result<InferenceResult, CoreError> {
        let session = self.require_session(session_id)?;
        let findings = infer_findings(&self.workspace.list_events(session_id)?);
        self.workspace.replace_findings(session_id, &findings)?;
        Ok(InferenceResult {
            session,
            findings: self.workspace.list_findings(session_id)?,
        })
    }

    pub fn compare_run(
        &self,
        session_a: &str,
        session_b: &str,
    ) -> Result<CompareResult, CoreError> {
        let a = self.timeline_get(session_a)?;
        let b = self.timeline_get(session_b)?;
        Ok(compare_sessions(&a, &b))
    }

    pub fn import_har(&self, session_id: &str, path: &Path) -> Result<ImportResult, CoreError> {
        self.require_session(session_id)?;
        let inputs = import_har(path).map_err(|_| CoreError::ImportInvalidFile)?;
        let stored = self
            .workspace
            .append_events(session_id, inputs)
            .map_err(|_| CoreError::ImportInvalidFile)?;
        Ok(import_result(stored))
    }

    pub fn import_procmon(&self, session_id: &str, path: &Path) -> Result<ImportResult, CoreError> {
        let date = today_utc()?;
        self.import_procmon_on_date(session_id, path, &date)
    }

    pub fn import_procmon_on_date(
        &self,
        session_id: &str,
        path: &Path,
        date: &str,
    ) -> Result<ImportResult, CoreError> {
        self.require_session(session_id)?;
        if date.trim().is_empty() {
            return Err(CoreError::InvalidArgument);
        }
        let inputs = import_procmon(path, date).map_err(|_| CoreError::ImportInvalidFile)?;
        let stored = self
            .workspace
            .append_events(session_id, inputs)
            .map_err(|_| CoreError::ImportInvalidFile)?;
        Ok(import_result(stored))
    }

    pub fn import_jsonl(
        &self,
        session_id: &str,
        path: &Path,
        map_path: &Path,
    ) -> Result<ImportResult, CoreError> {
        self.require_session(session_id)?;
        let inputs = import_jsonl(path, map_path).map_err(|_| CoreError::ImportInvalidFile)?;
        let stored = self
            .workspace
            .append_events(session_id, inputs)
            .map_err(|_| CoreError::ImportInvalidFile)?;
        Ok(import_result(stored))
    }

    pub fn launch_run(
        &self,
        session_id: &str,
        request: LaunchRequest,
    ) -> Result<LaunchResult, CoreError> {
        self.require_session(session_id)?;
        launch_target(&self.workspace, session_id, request).map_err(|error| match error {
            CoreError::InvalidArgument => CoreError::InvalidArgument,
            _ => CoreError::LaunchFailed,
        })
    }

    pub fn collector_append_event(
        &self,
        session_id: &str,
        input: TraceEventInput,
    ) -> Result<TraceEvent, CoreError> {
        self.require_session(session_id)?;
        self.workspace.append_event(session_id, input)
    }

    pub fn export_run(
        &self,
        session_id: &str,
        out_dir: &Path,
    ) -> Result<ServiceExportResult, CoreError> {
        self.require_session(session_id)?;
        let ExportResult { out_dir, files } = export_session(&self.workspace, session_id, out_dir)
            .map_err(|_| CoreError::ExportFailed)?;
        Ok(ServiceExportResult {
            cancelled: false,
            out_dir,
            files,
        })
    }
}
